### ISSUE TITLE
Segfault possibly caused by mysql2 0.4.0 (with test case)

### ISSUE THREAD

**Issue by saveriomiroddi** [brianmario/mysql2#1063]
Hello!

I'm getting a segfault under certain circumstances. I'm not sure what is the exact cause, but it may be related to `mysql2`.

Test conditions and issue reproducibility:

- Ruby 2.4.6, Rails 5.0.7,    mysql2 0.5.2 -> ok
- Ruby 2.4.6, Rails 4.2.11.1, mysql2 0.4.10 -> kaboom!
- Ruby 2.5.5, Rails 4.2.11.1, mysql2 0.4.10 -> kaboom!

My system is an Ubuntu 18.04-based distro. The db is MySQL 8.0.13 (defaults: `utf8m4`/`utf8mb4_0900_ai_ci`).

Test case, to be run with bash:

```bash
MYSQL_USER=
export MYSQL_PWD=
export MYSQL_SOCKET= # fill only if required

ruby -v # 2.4.6
gem install rails --version '~> 4.2.0'
rails new testjor
cd !$
bin/spring binstub --remove --all # rails new --skip-spring has no effect
echo "gem 'mysql2', '~> 0.4.0'" >> Gemfile
bundle install
bundle exec gem list '^(rails|mysql2)$' # 0.4.10/4.2.11.1
echo 'class MyModel < ActiveRecord::Base; end' > app/models/my_model.rb
mysql -u "$MYSQL_USER" -e 'CREATE SCHEMA testissue; CREATE TABLE testissue.my_models (id INT PRIMARY KEY AUTO_INCREMENT, myint INT, mystr VARCHAR(255))'
perl -i -pe 's/db\/development.sqlite3/testissue/' config/database.yml
perl -i -pe 's/(adapter:) sqlite3/$1 mysql2/' !$
[[ "$MYSQL_SOCKET" != "" ]] && perl -i -lpe '/adapter/ && print "  socket: $ENV{MYSQL_SOCKET}"' !$
bundle exec rails r "MyModel.count" # kaboom!
```

**Comment by sodabrew** [comment]
Yes, this is correct. There's no release of mysql2 0.4.x that supports the new MySQL 8 encodings.

Is the bug report is that a version of mysql2 older than the current release causes crashes, but the current release does not?

... there's been a few people asking for a new release of the 0.4.x branch. Gaaaah agh, time and effort, etc.

**Comment by saveriomiroddi** [comment]
> Is the bug report is that a version of mysql2 older than the current release causes crashes, but the current release does not?

If I understand correctly, yes: `mysql2` v0.5.2 (at least when paired with Rails 5.0.7) does _not_ cause the segfault.

**Comment by krisleech** [comment]
Firstly, thanks for the great work maintaining the mysql2 gem.

I ran in to this today after upgrading Ubuntu. The README seems to suggest that mysql2 version 0.4 works with MySQL 8.x.

The solution for me was simple:

```diff
-gem 'mysql2', '< 0.5' # [1]
+gem 'mysql2', '~>0.5.0' # [1]
```

(I have mysql2 pinned because ActiveRecord needs a specific version)

* Ubuntu 19.10
* MySQL 8.0.17
* mysql2 0.4.x segfaults
* mysql2 0.5.x works
* Rails 5.2.3

**Comment by krisleech** [comment]
[Setting the collation](https://stackoverflow.com/questions/58930916/how-to-create-migrations-with-specific-collate) might be of help to those using Rails and MySQL 8.
